rls/src/main/java/io/grpc/rls/internal/LbPolicyConfiguration.java

```diff
@@ -0,0 +1,446 @@
+        this(childHelperProvider, null);
+      }
+
+      ChildPolicyReportingHelper(
+          ChildLoadBalancerHelperProvider childHelperProvider,
+          @Nullable ChildLbStatusListener listener) {
+        checkNotNull(childHelperProvider, "childHelperProvider");
+        this.delegate = childHelperProvider.forTarget(getTarget());
+        this.listener = listener;
+      }
+
+      @Override
+      protected Helper delegate() {
+        return delegate;
+      }
+
+      @Override
+      public void updateBalancingState(ConnectivityState newState, SubchannelPicker newPicker) {
+        setPicker(newPicker);
+        super.updateBalancingState(newState, newPicker);
+        if (listener != null) {
+          listener.onStatusChanged(newState);
+        }
+      }
+
+      @Override
+      public Subchannel createSubchannel(CreateSubchannelArgs args) {
+        final Subchannel subchannel = super.createSubchannel(args);
+        return new ForwardingSubchannel() {
+          @Override
+          protected Subchannel delegate() {
+            return subchannel;
+          }
+
+          @Override
+          public void start(final SubchannelStateListener listener) {
+            super.start(new SubchannelStateListener() {
+              @Override
+              public void onSubchannelState(ConnectivityStateInfo newState) {
+                setConnectivityStateInfo(newState);
+                listener.onSubchannelState(newState);
+              }
+            });
+          }
+        };
+      }
+    }
+
+    /** Listener for child lb status change events. */
+    interface ChildLbStatusListener {
+
+      /** Notifies when child lb status changes. */
+      void onStatusChanged(ConnectivityState newState);
+    }
+  }
+
+  private static final class RefCountedChildPolicyWrapper
+      implements ObjectPool<ChildPolicyWrapper> {
+
+    private final AtomicLong refCnt = new AtomicLong(1);
+    @Nullable
+    private ChildPolicyWrapper childPolicyWrapper;
+
+    private RefCountedChildPolicyWrapper(ChildPolicyWrapper childPolicyWrapper) {
+      this.childPolicyWrapper = checkNotNull(childPolicyWrapper, "childPolicyWrapper");
+    }
+
+    @Override
+    public ChildPolicyWrapper getObject() {
+      long curr = refCnt.getAndIncrement();
+      if (curr <= 0) {
+        throw new IllegalStateException("ChildPolicyWrapper is already released");
+      }
+      return childPolicyWrapper;
+    }
+
+    @Override
+    @Nullable
+    public ChildPolicyWrapper returnObject(Object object) {
+      checkState(
+          childPolicyWrapper != null,
+          "cannot return already released ChildPolicyWrapper, this is possibly a bug.");
+      checkState(
+          childPolicyWrapper == object,
+          "returned object doesn't match the pooled childPolicyWrapper");
+      long newCnt = refCnt.decrementAndGet();
+      if (newCnt == 0) {
+        childPolicyWrapper = null;
+      }
+      return childPolicyWrapper;
```

Comment:
This API should always return null.
https://github.com/grpc/grpc-java/blob/v1.28.1/core/src/main/java/io/grpc/internal/ObjectPool.java#L32-L36

Reply:
Might need introduce something like `isReleased()`

Reply:
tried to avoid additional check, but this sounds better.